### PR TITLE
Update DataProcessor to track training and eval modes, add base DataProcessor class with required abstractmethods

### PR DESCRIPTION
**Pull request by dhpitt**
quick PR to address a structural fix to @Robertboy18's IFNO PR #274. Allows callbacks to modify the state of a data processor

### CODE REVIEW

File: neuralop/training/trainer.py [hunk 1]
```diff
@@ -112,21 +129,26 @@ def train(self, train_loader, test_loaders,
...
 
         if self.callbacks:
-            self.callbacks.on_train_start(train_loader=train_loader, test_loaders=test_loaders,
-                                    optimizer=optimizer, scheduler=scheduler, 
-                                    regularizer=regularizer, training_loss=training_loss, 
-                                    eval_losses=eval_losses)
-            
+            self.callbacks.on_train_start(
+                train_loader=train_loader,
+                test_loaders=test_loaders,
```
Comment (JeanKossaifi) [inline neuralop/training/trainer.py:134]: Are there some weird trailing spaces here?

File: neuralop/training/trainer.py [hunk 2]
```diff
@@ -44,29 +47,35 @@ def __init__(self, *,
         """
 
         if callbacks:
-            assert type(callbacks) == list, "Callbacks must be a list of Callback objects"
+            assert (
+                type(callbacks) == list
```
Comment (JeanKossaifi) [inline neuralop/training/trainer.py:51]: maybe just isinstance(callbacks, list)?

File: neuralop/training/trainer.py [hunk 3]
```diff
@@ -44,29 +47,35 @@ def __init__(self, *,
...
+            ), "Callbacks must be a list of Callback objects"
             self.callbacks = PipelineCallback(callbacks=callbacks)
-            self.override_load_to_device = (self.callbacks.device_load_callback_idx is not None)
+            self.override_load_to_device = (
```
Comment (JeanKossaifi) [inline neuralop/training/trainer.py:54]: Looks like some trailing spaces/tabs?

File: neuralop/training/trainer.py [hunk 4]
```diff
@@ -149,27 +172,35 @@ def train(self, train_loader, test_loaders,
                     sample = self.data_processor.preprocess(sample)
                 else:
                     # load data to device if no preprocessor exists
-                    sample = {k:v.to(self.device) for k,v in sample.items() if torch.is_tensor(v)}
+                    sample = {
```
Comment (JeanKossaifi) [inline neuralop/training/trainer.py:175]: Same, looks like strange spacing here

File: neuralop/training/trainer.py [hunk 5]
```diff
@@ -149,27 +172,35 @@ def train(self, train_loader, test_loaders,
...
                     self.callbacks.on_before_loss(out=out)
 
-                loss = 0.
+                loss = 0.0
 
                 if self.overrides_loss:
                     if isinstance(out, torch.Tensor):
-                        loss += self.callbacks.compute_training_loss(out=out.float(), **sample, amp_autocast=self.amp_autocast)
+                        loss += self.callbacks.compute_training_loss(
+                            out=out.float(), **sample, amp_autocast=self.amp_autocast
```
Comment (JeanKossaifi) [inline neuralop/training/trainer.py:198]: Why are we casting out to float?

Comment (JeanKossaifi) [reply neuralop/training/trainer.py:198]: What about low precision training?

File: neuralop/training/trainer.py [hunk 6]
```diff
@@ -149,27 +172,35 @@ def train(self, train_loader, test_loaders,
...
+                        )
                     elif isinstance(out, dict):
-                        loss += self.callbacks.compute_training_loss(**out, **sample, amp_autocast=self.amp_autocast)
+                        loss += self.callbacks.compute_training_loss(
+                            **out, **sample, amp_autocast=self.amp_autocast
```
Comment (JeanKossaifi) [inline neuralop/training/trainer.py:202]: I wonder if it is not confusing to have out in one case and **out, **sample in the other

**Review by JeanKossaifi — COMMENTED** [review]
Looks good to me overall - left some comments. The spacing looks odd in places.

File: neuralop/training/trainer.py (hunk 6 above)
Comment (dhpitt) [reply neuralop/training/trainer.py:207]: changed this, as we never implemented a dict

File: neuralop/training/trainer.py (hunk 3 above)
Comment (dhpitt) [reply neuralop/training/trainer.py:54]: re-autoformatted

**Comment by JeanKossaifi** [comment]
Thanks @dhpitt, merging!

**Pull request merged**